cli: Add Hubble section to cilium status output

gandro opened the pull request:
This adds a new line to the output of `cilium status` which prints the
state of the embedded Hubble server. If enabled, the output will look
similar to `hubble status`:

```console
$ cilium status
[...]
Hubble:   Ok   Current/Max Flows: 2542/4096 (62.06%), Flows/s: 164.21   Metrics: Disabled
```

glibsm commented on pkg/client/client.go:
```diff
@@ -412,4 +412,15 @@ func FormatStatusResponse(w io.Writer, sr *models.StatusResponse, allAddresses,
+	if sr.Hubble != nil {
+		h := sr.Hubble
+		msg := h.Msg
+		if h.State == models.HubbleStatusStateOk && msg == "" && h.MaxFlows > 0 {
```
if the status is not ok, what's wrong with printing the buffer contents?

grpc server failing, but buffer being populated is a valid data point i think

gandro replied:
The way the status is implemented right now, a non-empty `Msg` and us being able to obtain the buffer statistics are mutually exclusive. But I can change the client logic to be future proof in case this ever changes.

rolinh commented on daemon/cmd/daemon.go:
```diff
@@ -137,6 +138,8 @@ type Daemon struct {
 
 	k8sWatcher *watchers.K8sWatcher
+
+	hubbleServer *hubbleServer.LocalObserverServer
```
Would you mind giving this struct member a more explicit name such as `hubbleObserver` or so? As it, it might be confused with hubble server (the gRPC server currently in hubble's `cmd/serve`).

soumynathan replied:
+1

rolinh commented on daemon/cmd/hubble.go:
```diff
@@ -50,21 +80,21 @@ func (d *Daemon) launchHubble() {
+	go d.hubbleServer.Start()
+	d.monitorAgent.GetMonitor().RegisterNewListener(d.ctx, hubble.NewHubbleListener(d.hubbleServer))
 
 	srv, err := hubbleServe.NewServer(logger,
```
This is why I think it's confusing: it looks like you create a new hubble server with a hubble server as an option (the observer service).

soumynathan reviewed (commented):
LGTM

soumynathan reviewed (commented):
LGTM

gandro committed:
api/observer: Add total number of flows and uptime to Hubble status

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
hubble/server: Populate total number of flows and uptime in status

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
api/healthz: Introduce HubbleStatus object

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
daemon: Add status probe for Hubble observer

This adds a status probe which polls Hubble's ServerStatus RPC call
locally to obtain the current state of Hubble. Since the RPC stub is
invoked directly and not via a TCP or unix domain socket, this does
not replace a connectivity check.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
client: Add Hubble section to status output

This adds a new line to the output of `cilium status` which prints the
state of the embedded Hubble server. If enabled, the output will look
similar to `hubble status`:

```console
$ cilium status
[...]
Hubble:                 Ok   Current/Max Flows: 2542/4096 (62.06%), Flows/s: 164.21   Metrics: Disabled
```

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

rolinh commented on daemon/cmd/hubble.go:
```diff
@@ -15,8 +15,14 @@
+	"github.com/go-openapi/strfmt"
+
+	"github.com/cilium/cilium/api/v1/models"
+	"github.com/cilium/cilium/api/v1/observer"
```
Would you mind renaming this import to `observerpb`? cf #10892

rolinh commented on pkg/hubble/server/local_observer.go:
```diff
@@ -106,13 +112,15 @@ func NewLocalServer(
+		eventschan:       make(chan *observer.GetFlowsResponse, 100), // option here?
+		payloadParser:    payloadParser,
+		startTime:        time.Now(),
+		numObservedFlows: 0,
```
I don't think this member needs to be explicitly initialized as it's zero-value is 0.

rolinh commented on pkg/hubble/server/local_observer.go:
```diff
@@ -143,6 +151,7 @@ nextEvent:
 			}
 		}
 
+		s.numObservedFlows++
```
Since you increment this counter here, the reported observed flows might be different than what was actually sent to the ring buffer. Is this what we want? Do we want a "processed flows" variable to differentiate the two?

gandro replied:
That's a good point and something that actually already bit me, since this also includes events currently ignored by Hubble (e.g. agent notifications). I will move it.

aanm commented on pkg/client/client.go:
```diff
@@ -412,4 +412,35 @@ func FormatStatusResponse(w io.Writer, sr *models.StatusResponse, allAddresses,
+		}
+
+		fmt.Fprintf(w, "Hubble:\t%s\n", strings.Join(fields, "\t"))
+	}
```
Should we have a branch that prints "Hubble: Disabled"?

gandro replied:
So if the local cilium-agent has this PR applied, it should print "Hubble: Disabled", since the server always populates `sr.Hubble`. The only case where it would not print a `Hubble: Disabled` line is when you use this client with an older version of cilium-agent which does have the `HubbleStatus` in `/healthz`.

Considering all many of the other branches in this file also don't print anything if their corresponding section is missing, I think it's better to keep it as.

aanm reviewed (approved):
LGTM, only a small nit.
